client/utils/src/pubsub.rs

```diff
@@ -0,0 +1,298 @@
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Provides means to implement a typical Pub/Sub mechanism.
+//!
+//! This module provides a type `Hub` which can be used both to subscribe,
+//! and to send the broadcast messages.
+//!
+//! The `Hub` type is parametrized by two other types:
+//! - `Channel` — operates with the underlying channels;
+//! - `Registry` — implementation of the subscription/dispatch logic.
+//!
+//! A Registry is implemented by defining the following traits:
+//! - `Subscribe<K>`;
+//! - `Dispatch<M>`;
+//! - `Unsubscribe`.
+//!
+//! As a result of subscription `Hub::subscribe` method returns an instance of `Receiver<Channel,
+//! Registry>`. That can be used as a `futures::Stream` to receive the messages.
+//! Upon drop the `Receiver<Channel, Registry>` shall unregister itself from the `Hub`.
+
+use std::{
+	collections::HashMap,
+	ops::DerefMut,
+	pin::Pin,
+	sync::{Arc, Weak},
+	task::{Context, Poll},
+};
+
+use ::futures::stream::{FusedStream, Stream};
+use ::parking_lot::Mutex;
+
+pub mod channels;
+
+/// The type to identify subscribers.
+pub type SubsID = crate::id_sequence::SeqID;
+
+/// Unsubscribe: unregisters a previously created subscription.
+pub trait Unsubscribe {
+	/// Remove all registrations of the subscriber with ID `subs_id`.
+	fn unsubscribe(&mut self, subs_id: &SubsID);
+}
+
+/// Subscribe using a key of type `K`
+pub trait Subscribe<K> {
+	/// Register subscriber with the ID `subs_id` as having interest to the key `K`.
+	fn subscribe(&mut self, subs_key: K, subs_id: SubsID);
+}
+
+/// Dispatch a message of type `M`.
+pub trait Dispatch<M> {
+	/// The type of the that shall be sent through the channel as a result of such dispatch.
+	type Item;
+
+	/// Dispatch the message of type `M`.
+	///
+	/// The implementation is given an instance of `M` and is supposed to invoke `dispatch` for
+	/// each matching subscriber, with an argument of type `Self::Item` matching that subscriber.
+	///
+	/// Note that this does not have to be of the same type with the item that will be sent through
+	/// to the subscribers. The subscribers will receive a message of type `Self::Item`.
+	fn dispatch<F>(&mut self, message: M, dispatch: F)
+	where
+		F: FnMut(&SubsID, Self::Item);
+}
+
+/// Channel routines.
+///
+/// Allows to create a pair of tx and rx, and to send a message over the tx.
+pub trait Channel {
+	/// The sending side of the channel.
+	type Tx;
+
+	/// The receiving side of the channel.
+	type Rx;
+
+	/// The type of an item that can be sent through this channel.
```

Comment:
Why do we need this?

We currently only have one implementation. Where is the need to have this that generic?

Reply:
I understand that currently the decision on using the unbounded channels has been made.
But if any other policy for handling slow consumers is to be introduced, that would likely to be tried out first on some isolated area: in that case we would just need to implement another type with `Channel` trait defined for it.

Another reason — to have the channel-shuffling code separate from the other logic. It just seemed more neat for me this way. 

If you insist, I will flatten that, so that this trait won't be introduced.

Reply:
The plan is to get rid of these unbounded channels, but they will completely disappear. Currently they are just hacks to combine async/sync code. 

Please remove these extra traits etc.